### PR TITLE
Plan publishable crates before publish

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- add a publish planning dataclass that filters crates based on manifest publish flags and configuration excludes
- surface the publish plan through the CLI output, behavioural tests, and refreshed documentation
- replace the placeholder publish unit tests with coverage for the new planner and update the roadmap

## Testing
- make check-fmt
- make typecheck
- make lint
- make test

------
https://chatgpt.com/codex/tasks/task_e_68f62fc6b8c883229ed9ccadb82a5411

## Summary by Sourcery

Implement crate publication planning in the `publish` command by filtering on manifest and configuration excludes, rendering a structured plan in CLI output, and updating docs, Makefile and tests to reflect the new behavior.

New Features:
- Add PublishPlan dataclass and plan_publication function to compute publishable, skipped, and missing crates

Enhancements:
- Replace placeholder publish command with real planning logic and integrate configuration/workspace loading in run()
- Update Makefile lint/format/typecheck targets and adjust ruff config rules

Documentation:
- Update usage guide, design doc, roadmap, and AGENTS.md to describe new publish planning feature and revised testing commands

Tests:
- Introduce unit tests for publish planner and run() behavior, plus BDD fixtures and step definitions for publish command scenarios

## Summary by CodeRabbit

* **New Features**
  * Publish now generates and displays a structured publication plan up front, showing crates to publish, crates skipped (by manifest or configuration), and unmatched/stale exclusions.

* **Documentation**
  * Usage, design, and roadmap updated with publish plan examples, semantics, and testing guidance.

* **Tests**
  * Added unit tests and extensive BDD fixtures/steps for publish and bump flows; removed legacy placeholder test module and obsolete step implementations.

* **Chores**
  * Build/tooling targets standardized and lint config adjusted.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[lading/commands/publish.py](https://github.com/leynos/crate-tools/pull/16/files/4a9ddfb0bce907633f2abaac3fa75b9788cb8aaa#diff-603138d2844fc0dbc4eeda587b905cfbb8a58ddd00adaf36ac0027cabc981cde)

Comment on lines +73 to +100

~~~python
def _format_plan(
    plan: PublishPlan, *, strip_patches: config_module.StripPatchesSetting
) -> str:
    """Render ``plan`` to a human-readable summary for CLI output."""
    lines = [
        f"Publish plan for {plan.workspace_root}",
        f"Strip patch strategy: {strip_patches}",
    ]

    if plan.publishable:
        lines.append(f"Crates to publish ({len(plan.publishable)}):")
        lines.extend(f"- {crate.name} @ {crate.version}" for crate in plan.publishable)
    else:
        lines.append("Crates to publish: none")

    if plan.skipped_manifest:
        lines.append("Skipped (publish = false):")
        lines.extend(f"- {crate.name}" for crate in plan.skipped_manifest)

    if plan.skipped_configuration:
        lines.append("Skipped via publish.exclude:")
        lines.extend(f"- {crate.name}" for crate in plan.skipped_configuration)

    if plan.missing_configuration_exclusions:
        lines.append("Configured exclusions not found in workspace:")
        lines.extend(f"- {name}" for name in plan.missing_configuration_exclusions)

    return "\n".join(lines)
~~~

❌ New issue: [Complex Method](https://codescene.io/projects/71839/delta?repo-id=1205555&review-id=16&biomarker=Complex+Method&filename=lading%2Fcommands%2Fpublish.py&method=_format_plan)
_format_plan has a cyclomatic complexity of 10, threshold = 9

File: tests/unit/test_publish_command.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+def test_plan_publication_records_missing_exclusions(tmp_path: Path) -> None:
+    """Unknown entries in publish.exclude are reported in the plan."""
+    root = tmp_path.resolve()
+    workspace = _make_workspace(root)
+    configuration = _make_config(exclude=("missing",))
+
+    plan = publish.plan_publication(workspace, configuration)
+
+    assert plan.missing_configuration_exclusions == ("missing",)
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/unit/test_publish_command.py]: Has this now been resolved?

File: tests/bdd/steps/test_cli_steps.py [hunk 2]
```diff
@@ -547,14 +547,15 @@ def _then_dependency_requirement_step(
...
+    """Assert that the publish command emits a publication plan summary."""
     assert cli_run["returncode"] == 0
     workspace = cli_run["workspace"]
-    stdout = cli_run["stdout"]
-    assert f"publish placeholder invoked for {workspace}" in stdout
-    assert "(crates: 1 crate, strip patches: all)" in stdout
+    lines = [line.strip() for line in cli_run["stdout"].splitlines() if line.strip()]
+    assert lines[0] == f"Publish plan for {workspace}"
+    assert "Strip patch strategy: all" in lines[1]
+    assert any(line == f"- {crate_name} @ 0.1.0" for line in lines)
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/bdd/steps/test_cli_steps.py]: Has this now been resolved?

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[tests/bdd/steps/fixtures.py](https://github.com/leynos/crate-tools/pull/16/files/4efbd7f9cdf1d25f7ebd310e9c7712a5e8734cff#diff-982242af73ff0a8bfbf42316d285e5b7d34b12899f97c81c5fce969a34dc88f7)

Comment on file

~~~python
    version: str,
@given(parsers.parse('publish.exclude contains "{crate_name}"'))
def given_publish_exclude_contains(
    workspace_directory: Path,
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71839/delta?repo-id=1205555&review-id=16&biomarker=Code+Duplication&filename=tests%2Fbdd%2Fsteps%2Ffixtures.py&method=)
The module contains 2 functions with similar structure: given_bump_exclude_contains,given_publish_exclude_contains

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[lading/commands/publish.py](https://github.com/leynos/crate-tools/pull/16/files/1efbd6132c8b57fc4849caec09c67c631d6b6d5f#diff-603138d2844fc0dbc4eeda587b905cfbb8a58ddd00adaf36ac0027cabc981cde)

Comment on file

~~~python
    workspace: WorkspaceGraph | None = None,
) -> str:
    """Return a placeholder message for the publish command."""
    """Plan crate publication for ``workspace_root``."""
~~~

❌ New issue: [Bumpy Road Ahead](https://codescene.io/projects/71839/delta?repo-id=1205555&review-id=16&biomarker=Bumpy+Road+Ahead&filename=lading%2Fcommands%2Fpublish.py&method=run)
run has 2 blocks with nested conditional logic. Any nesting of 2 or deeper is considered. Threshold is 2 blocks per function

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

- Consider sorting the publishable and skipped crate lists alphabetically to ensure stable, reproducible CLI output regardless of the workspace graph’s internal ordering.
- There’s a lot of similar logic across the `_format_*_section` functions; you could extract a small helper that emits an optional heading plus a list of items to reduce boilerplate.

**Comment by leynos** [comment]
@sourcery-ai review

File: Makefile [hunk 3]
```diff
@@ -73,8 +73,7 @@ markdownlint: $(MDLINT) ## Lint Markdown files
 	  -not -path './.venv/*' -print0 | xargs -0 $(MDLINT)
 
 nixie: $(NIXIE) ## Validate Mermaid diagrams
-	find . -type f -name '*.md' \
-	  -not -path './.venv/*' -print0 | xargs -0 $(NIXIE)
+	nixie --no-sandbox
```
Comment (SourceryAI) [inline Makefile]: **🚨 question (security):** Switching to nixie --no-sandbox may reduce validation isolation.

Disabling sandboxing may expose the validation process to security risks if files are untrusted. Confirm that this is necessary and that all inputs are safe.

File: tests/bdd/steps/test_publish_steps.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    workspace = cli_run["workspace"]
+    lines = [line.strip() for line in cli_run["stdout"].splitlines() if line.strip()]
+    assert lines[0] == f"Publish plan for {workspace}"
+    assert "Strip patch strategy: all" in lines[1]
+    assert f"- {crate_name} @ 0.1.0" in lines
+
+
+def _publish_plan_lines(cli_run: dict[str, typ.Any]) -> list[str]:
+    """Return trimmed publish plan output lines for ``cli_run``."""
+    return [line.strip() for line in cli_run["stdout"].splitlines() if line.strip()]
```
Comment (SourceryAI) [inline tests/bdd/steps/test_publish_steps.py]: **suggestion (testing):** Consider adding a BDD step for the case where no crates are publishable.

Please add a BDD scenario to verify the CLI output when no crates are publishable, ensuring this edge case is handled clearly.

Suggested implementation:

```python
def then_publish_prints_plan(cli_run: dict[str, typ.Any], crate_name: str) -> None:
    """Assert that the publish command emits a publication plan summary."""
    assert cli_run["returncode"] == 0
    workspace = cli_run["workspace"]
    lines = [line.strip() for line in cli_run["stdout"].splitlines() if line.strip()]
    assert lines[0] == f"Publish plan for {workspace}"
    assert "Strip patch strategy: all" in lines[1]
    assert f"- {crate_name} @ 0.1.0" in lines


@then(parsers.parse('no crates are publishable'))
def then_publish_prints_no_crates_publishable(cli_run: dict[str, typ.Any]) -> None:
    """Assert that the publish command emits a message when no crates are publishable."""
    assert cli_run["returncode"] == 0
    lines = [line.strip() for line in cli_run["stdout"].splitlines() if line.strip()]
    # Adjust the expected message to match your CLI's actual output
    assert any(
        "No crates are publishable" in line or "No crates to publish" in line
        for line in lines
    ), f"Expected a message indicating no crates are publishable, got: {lines}"

```

You will need to add a corresponding BDD scenario in your feature file (e.g., `publish.feature`) that uses the step: `Then no crates are publishable`.
Make sure your CLI emits a clear message like "No crates are publishable" or "No crates to publish" when this edge case occurs.
Adjust the expected message in the assertion if your CLI uses different wording.

**Review by SourceryAI — COMMENTED** [review]
Hey there - I've reviewed your changes and found some issues that need to be addressed.

<details>
<summary>Prompt for AI Agents</summary>

~~~markdown
Please address the comments from this code review:

## Individual Comments

### Comment 1
<location> `Makefile:76` </location>
<code_context>
 nixie: $(NIXIE) ## Validate Mermaid diagrams
-	find . -type f -name '*.md' \
-	  -not -path './.venv/*' -print0 | xargs -0 $(NIXIE)
+	nixie --no-sandbox

 test: build uv pytest ## Run tests
</code_context>

<issue_to_address>
**🚨 question (security):** Switching to nixie --no-sandbox may reduce validation isolation.

Disabling sandboxing may expose the validation process to security risks if files are untrusted. Confirm that this is necessary and that all inputs are safe.
</issue_to_address>

### Comment 2
<location> `tests/bdd/steps/test_publish_steps.py:39-41` </location>
<code_context>
+    assert f"- {crate_name} @ 0.1.0" in lines
+
+
+def _publish_plan_lines(cli_run: dict[str, typ.Any]) -> list[str]:
+    """Return trimmed publish plan output lines for ``cli_run``."""
+    return [line.strip() for line in cli_run["stdout"].splitlines() if line.strip()]
+
+
</code_context>

<issue_to_address>
**suggestion (testing):** Consider adding a BDD step for the case where no crates are publishable.

Please add a BDD scenario to verify the CLI output when no crates are publishable, ensuring this edge case is handled clearly.

Suggested implementation:

```python
def then_publish_prints_plan(cli_run: dict[str, typ.Any], crate_name: str) -> None:
    """Assert that the publish command emits a publication plan summary."""
    assert cli_run["returncode"] == 0
    workspace = cli_run["workspace"]
    lines = [line.strip() for line in cli_run["stdout"].splitlines() if line.strip()]
    assert lines[0] == f"Publish plan for {workspace}"
    assert "Strip patch strategy: all" in lines[1]
    assert f"- {crate_name} @ 0.1.0" in lines


@then(parsers.parse('no crates are publishable'))
def then_publish_prints_no_crates_publishable(cli_run: dict[str, typ.Any]) -> None:
    """Assert that the publish command emits a message when no crates are publishable."""
    assert cli_run["returncode"] == 0
    lines = [line.strip() for line in cli_run["stdout"].splitlines() if line.strip()]
    # Adjust the expected message to match your CLI's actual output
    assert any(
        "No crates are publishable" in line or "No crates to publish" in line
        for line in lines
    ), f"Expected a message indicating no crates are publishable, got: {lines}"

```

You will need to add a corresponding BDD scenario in your feature file (e.g., `publish.feature`) that uses the step: `Then no crates are publishable`.
Make sure your CLI emits a clear message like "No crates are publishable" or "No crates to publish" when this edge case occurs.
Adjust the expected message in the assertion if your CLI uses different wording.
</issue_to_address>

### Comment 3
<location> `lading/commands/publish.py:115` </location>
<code_context>
+    return [header, *formatted_items]
+
+
+def _format_plan(
+    plan: PublishPlan, *, strip_patches: config_module.StripPatchesSetting
+) -> str:
</code_context>

<issue_to_address>
**issue (complexity):** Consider inlining section formatting in _format_plan and removing helper indirection in run for improved clarity.

```suggestion
# Drop `_format_section` and inline per‐section formatting in `_format_plan`
# (also removes the generic `T` and `empty_lines` hack)

-def _format_plan(
-    plan: PublishPlan, *, strip_patches: config_module.StripPatchesSetting
-) -> str:
-    lines = [
-        f"Publish plan for {plan.workspace_root}",
-        f"Strip patch strategy: {strip_patches}",
-    ]
-
-    lines.extend(
-        _format_section(
-            plan.publishable,
-            header=f"Crates to publish ({len(plan.publishable)}):",
-            item_formatter=lambda crate: f"- {crate.name} @ {crate.version}",
-            empty_lines=("Crates to publish: none",),
-        )
-    )
-    # … other sections …
-    return "\n".join(lines)
+def _format_plan(
+    plan: PublishPlan,
+    *,
+    strip_patches: config_module.StripPatchesSetting
+) -> str:
+    lines: list[str] = [
+        f"Publish plan for {plan.workspace_root}",
+        f"Strip patch strategy: {strip_patches}",
+    ]
+
+    # Crates to publish
+    if plan.publishable:
+        lines.append(f"Crates to publish ({len(plan.publishable)}):")
+        lines.extend(f"- {c.name} @ {c.version}" for c in plan.publishable)
+    else:
+        lines.append("Crates to publish: none")
+
+    # Skipped (publish = false)
+    if plan.skipped_manifest:
+        lines.append("Skipped (publish = false):")
+        lines.extend(f"- {c.name}" for c in plan.skipped_manifest)
+
+    # Skipped via configuration
+    if plan.skipped_configuration:
+        lines.append("Skipped via publish.exclude:")
+        lines.extend(f"- {c.name}" for c in plan.skipped_configuration)
+
+    # Missing exclusions
+    if plan.missing_configuration_exclusions:
+        lines.append("Configured exclusions not found in workspace:")
+        lines.extend(f"- {name}" for name in plan.missing_configuration_exclusions)
+
+    return "\n".join(lines)
```

```suggestion
# Inline the two `_ensure_*` helpers directly in `run` to remove indirection

-def run(
-    workspace_root: Path,
-    configuration: LadingConfig | None = None,
-    workspace: WorkspaceGraph | None = None,
-) -> str:
-    root = normalise_workspace_root(workspace_root)
-    active_configuration = _ensure_configuration(configuration, root)
-    active_workspace = _ensure_workspace(workspace, root)
-    plan = plan_publication(active_workspace, active_configuration, workspace_root=root)
-    return _format_plan(plan, strip_patches=active_configuration.publish.strip_patches)
+def run(
+    workspace_root: Path,
+    configuration: LadingConfig | None = None,
+    workspace: WorkspaceGraph | None = None,
+) -> str:
+    root = normalise_workspace_root(workspace_root)
+
+    # load or fetch configuration
+    if configuration is None:
+        try:
+            configuration = config_module.current_configuration()
+        except config_module.ConfigurationNotLoadedError:
+            configuration = config_module.load_configuration(root)
+
+    # load or reuse workspace
+    if workspace is None:
+        from lading.workspace import load_workspace, WorkspaceModelError
+        try:
+            workspace = load_workspace(root)
+        except FileNotFoundError as e:
+            raise WorkspaceModelError(f"Workspace root not found: {root}") from e
+
+    plan = plan_publication(workspace, configuration, workspace_root=root)
+    return _format_plan(plan, strip_patches=configuration.publish.strip_patches)
```
</issue_to_address>

### Comment 4
<location> `lading/commands/publish.py:36` </location>
<code_context>
+        return tuple(crate.name for crate in self.publishable)
+
+
+def plan_publication(
+    workspace: WorkspaceGraph,
+    configuration: LadingConfig,
</code_context>

<issue_to_address>
**issue (review_instructions):** Add unit and behavioural tests for the new publication planning logic.

The new functions and logic for publication planning (e.g., plan_publication, PublishPlan, _format_plan, etc.) require both unit and behavioural tests to verify correct behaviour and edge cases. No new tests are present in the diff.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*`

**Instructions:**
For any new feature or change to an existing feature, both behavioural *and* unit tests are required.

</details>
</issue_to_address>

### Comment 5
<location> `lading/commands/publish.py:186` </location>
<code_context>
+        raise WorkspaceModelError(message) from exc


 def run(
</code_context>

<issue_to_address>
**issue (review_instructions):** Add tests for the new run function implementation.

The run function was substantially changed to implement publication planning. This new behaviour must be covered by both unit and behavioural tests, but no new tests are present in the diff.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*`

**Instructions:**
For any new feature or change to an existing feature, both behavioural *and* unit tests are required.

</details>
</issue_to_address>

### Comment 6
<location> `docs/usage-guide.md:127` </location>
<code_context>
-count, and returns successfully. Publication planning and execution will arrive
-in later phases of the roadmap.
+`publish` now produces a publication plan for the workspace. The command reads
+`publish.exclude` from `lading.toml`, honours any crate manifests that declare
+`publish = false`, and prints a structured summary listing the crates that will
+be published. Additional sections document crates skipped by manifest flags or
</code_context>

<issue_to_address>
**suggestion (review_instructions):** The word 'honours' uses en-GB spelling; en-oxendic spelling is preferred ('honors').

Please update 'honours' to 'honors' to match the en-oxendic spelling convention specified in the review instructions.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Use en-oxendic (-ize / -yse / -our) spelling and grammar.

</details>
</issue_to_address>

### Comment 7
<location> `docs/usage-guide.md:148` </location>
<code_context>
+```
+
+When the configuration excludes additional crates, or a manifest sets
+`publish = false`, the plan prints dedicated sections so the operator can see
+why those crates were skipped.
+
</code_context>

<issue_to_address>
**suggestion (review_instructions):** The phrase 'so the operator can see why those crates were skipped' uses 2nd person construction; rephrase to avoid this.

Consider rephrasing to 'so the reasons for skipping crates are visible to the operator.'

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Avoid 2nd person or 1st person pronouns ("I", "you", "we")

</details>
</issue_to_address>

### Comment 8
<location> `tests/bdd/steps/fixtures.py:402` </location>
<code_context>
                publish=False if not publishable else None,

</code_context>

<issue_to_address>
**suggestion (code-quality):** Swap if/else branches of if expression to remove negation ([`swap-if-expression`](https://docs.sourcery.ai/Reference/Rules-and-In-Line-Suggestions/Python/Default-Rules/swap-if-expression))

```suggestion
                publish=None if publishable else False,
```

<br/><details><summary>Explanation</summary>Negated conditions are more difficult to read than positive ones, so it is best
to avoid them where we can. By swapping the `if` and `else` conditions around we
can invert the condition and make it positive.
</details>
</issue_to_address>
~~~

</details>

***

Hi @leynos! 👋

Thanks for trying out Sourcery by commenting with `@sourcery-ai review`! 🚀

Install the [sourcery-ai bot](https://github.com/apps/sourcery-ai/installations/new) to get automatic code reviews on every pull request ✨

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment and I'll use the feedback to improve your reviews.
</sub>

File: docs/usage-guide.md [hunk 5]
```diff
@@ -123,15 +123,31 @@ Dry run; would update version to 1.2.3 in 3 manifest(s):
 
 ### `publish`
 
-`publish` is scaffolded in the same fashion. It acknowledges the workspace,
-reports the configured `strip_patches` strategy, includes the discovered crate
-count, and returns successfully. Publication planning and execution will arrive
-in later phases of the roadmap.
+`publish` now produces a publication plan for the workspace. The command reads
+`publish.exclude` from `lading.toml`, honours any crate manifests that declare
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/usage-guide.md]: en-GB-oxendict spelling is "honours"

**Comment by leynos** [comment]
@sourcery-ai resolve
